cloud_blobstore/__init__.py

```diff
@@ -1,6 +1,65 @@
 import typing
 
 
+class PagedIter(typing.Iterable[str]):
+    """
+    Provide an iterator that will iterate over every object, filtered by prefix and delimiter. Alternately continue
+    iteration with token and key (start_after_key).
+    """
+
+    def get_api_response(self, next_token):
+        """
+        Make blobstore-specific list api request.
+        """
+        raise NotImplementedError()
+
+    def get_listing_from_response(self, resp) -> typing.Iterable[str]:
+        """
+        Retrieve blob key listing from blobstore response.
+        """
+        raise NotImplementedError()
+
+    def get_next_token_from_response(self, resp) -> str:
+        """
+        Retrieve opaque continuation token from blobstore response.
+        """
+        raise NotImplementedError()
+
+    def __iter__(self):
+        """
+        Iterate over the blobs, saving page tokens and blob key start_after_keys as needed in order to continue
+        listing where one left off.
+
+        If start_after_key is not None, iteration will begin on the next key if start_after_key is found on the
+        first page of results. If it is not found on the first page of results, BlobPagingError will be raised.
+        """
+        next_token = self.token
+
+        while True:
+            self.token = next_token
+
+            resp = self.get_api_response(next_token)
+            listing = self.get_listing_from_response(resp)
+
+            if self.start_after_key:
+                for key in listing:
```

Comment:
The biggest concern I have about this is that this can fail in a pretty silent way if `listing` is a sequence and not an iterable.  Syntactically, it's not an error, so python wouldn't throw an exception, but you would get the wrong results.

Reply:
Wow, good catch. I would just do `listing=iter(…)` above. IIRC, Python requires that  `i is iter(i)` for any iterator `i`:

```
>>> c=[]
>>> i = iter(c)
>>> i is iter(i)
True
```

Alternatively, unroll the inner loop and use `if key >= start_after_key` to skip elements already seen.

Reply:
That's fair. However, could-blobstore consumers should probably not be subclassing `PagedIter`. Perhaps this can be solved by moving it out of __init__.py so it's not available at package level?

Reply:
I don't understand this comment.  The issue is whether `get_listing_from_response(..)` can ever return a sequence instead of an iterator.

Reply:
Ok, interpreting your comments as criticisms of code maintainability, using next() instead of for .. in ..